jextract: 2023-04-14 -> 2023-11-27 (JDK 21)

msgilligan opened the pull request:
## Description of changes

Upgrade to build with JDK 21 and use the ~~2023-09-25~~ 2023-11-27 commit from the `jdk21` branch of `jextract`. See Issue #270791.

## Things done

~~Nothing except the changes~~. I am new to Nix ~~and don't even know how to test this locally~~. Submitting ~~as a DRAFT PR~~ to attach to Issue #270791.

**Update:** I've installed and tested with a small `jextract`/`j.l.foreign` project on `darwin-aarch64` and it seems to be working.

msgilligan reviewed (commented):
OK, I've tested this on aarch64-darwin and it seems to be working.

msgilligan commented:
> Yes, you should squash it first. I did a basic testing on my x86_64 machine and it seems OK. No other change is required if your testing on darwin goes successfully.

OK, I squashed and force-pushed. I also did some testing on `aarch64-darwin` and it seems to be working.

msgilligan commented:
@SharzyL is there anything else we need to do before this can be merged?

msgilligan commented:
> I think it is OK now, but I have no merge permission

So there is nothing else we need to do? Just wait for someone with merge permission to do the merge? Is there a way to notify them or assign the issue to them somehow?

jlesquembre commented on pkgs/development/tools/java/jextract/default.nix:
```diff
@@ -38,13 +38,13 @@ in
     repo = "jextract";
-    rev = "cf3afe9ca71592c8ebb32f219707285dd1d5b28a";
-    hash = "sha256-8qRD1Xg39vxtFAdguD8XvkQ8u7YzFU55MhyyJozVffo=";
+    rev = "a3c3644580a5144840de341e7c1cbf7dc4de3206"; # Tweak build for JDK 21
+    hash = "sha256-0Z8IskIF9GT3Y+Pah2G+TvCVGBK6++/1ctirv8ZsGBA=";
```
Could we use the latest JDK21 commit? I tested it and looks good: 

```suggestion
    rev = "f8d229775e8260d4bebf21525b3151b9fff128ae";
    hash = "sha256-jkUCh4oSgilszvvU7RpozFATIghaA9rJRAdUIl5jTHM=";
```

msgilligan replied:
I chose `a3c3644580a5144840de341e7c1cbf7dc4de3206` (`# Tweak build for JDK 21`) because it corresponds with the `jextract` binary released [on jdk.java.net](https://jdk.java.net/jextract/).

The hash you are suggesting (https://github.com/openjdk/jextract/commit/f8d229775e8260d4bebf21525b3151b9fff128ae) has a fix for a Windows test and a fix for one of the code samples so does not affect the `jextract` binary in any way.

There is also this commit https://github.com/openjdk/jextract/commit/8730fcf05c229d035b0db52ee6bd82622e9d03e9 (on `master`) that seems to have some actual fixes.

I'd be happy with any of those three commits. @SharzyL what do you suggest?

jlesquembre replied:
Since jextract doesn't have any releases yet, I'd go with https://github.com/openjdk/jextract/commit/8730fcf05c229d035b0db52ee6bd82622e9d03e9 since it included a fix. But that's my personal opinion.

SharzyL replied:
If you find these new changes useful, it is totally fine to include these new changes. Just remember to update the commit message, PR title and do tests. jextract have no release tags hence we can and only can pick commits from master.

msgilligan replied:
OK, I've updated to the current `master` of https://github.com/openjdk/jextract and tested on `aarch64-darwin` and it is working!

msgilligan committed:
jextract 2023-04-14 -> 2023-11-27 (JDK 21)

This updates jextract to use JDK 21 and to be be
based on the commit described as 
"Update jextract 21 with latest fixes" and as of
2023-12-12 is the latest commit on `master`.

Co-authored-by: Sharzy <me@sharzy.in>

SpriteOvO reviewed (approved):
LGTM, thanks!

jlesquembre commented:
Result of `nixpkgs-review pr 271127` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>jextract</li>
  </ul>
</details>
